### PR TITLE
use allreduce operation to sync trainers

### PR DESCRIPTION
**Pull request by sandyhouse**
use allreduce operation to sync trainers for some tasks such as downloading.

Todo:
   maybe run a global startup once to avoid generating a new ring at each startup program running.

### CODE REVIEW

File: plsc/entry.py [hunk 1]
```diff
@@ -550,8 +550,11 @@ def load_checkpoint(self,
 
             # sync all trainers to avoid loading checkpoints before 
             # parameters are downloaded
-            file_name = os.path.join(checkpoint_dir, '.lock')
+            file_name = os.path.join(checkpoint_dir, '.download_lock')
             if self.trainer_id == 0:
+                if os.path.exists(file_name):
+                    logger.error("Why the file {} exists?".format(file_name))
+                    exit()
```
Comment (gavin1332) [inline plsc/entry.py]: use allreduce to synchronize

Comment (sandyhouse) [reply plsc/entry.py]: done.

File: plsc/entry.py [hunk 2]
```diff
@@ -532,6 +532,39 @@ def _append_broadcast_ops(self, program):
                 outputs={'Out': var},
                 attrs={'use_calc_stream': True})
 
+    def _sync_program(self):
```
Comment (gavin1332) [inline plsc/entry.py]: `def _barrier(self):` may be better?

File: plsc/entry.py [hunk 3]
```diff
@@ -532,6 +532,39 @@ def _append_broadcast_ops(self, program):
...
+        """ Use allreduce operation to sync all trainers. """
```
Comment (gavin1332) [inline plsc/entry.py]: too complicated. Could a single program with allreduce or broadcast be enough?

File: plsc/entry.py (hunk 2 above)
Comment (sandyhouse) [reply plsc/entry.py]: done.

File: plsc/entry.py (hunk 3 above)
Comment (sandyhouse) [reply plsc/entry.py]: done.

File: plsc/entry.py [hunk 4]
```diff
@@ -532,6 +532,27 @@ def _append_broadcast_ops(self, program):
...
+        with fluid.program_guard(main_prog, startup_prog):
+            sync_data = fluid.layers.create_parameter(shape=[1],
+                                                      dtype="float32")
+            out = fluid.layers.collective._c_allreduce(sync_data,
+                                                       use_calc_stream=True)
+
+        gpu_id = int(os.getenv("FLAGS_selected_gpus", 0))
+        place = fluid.CUDAPlace(gpu_id)
+        exe = fluid.Executor(place)
+        exe.run(startup_prog)
```
Comment (gavin1332) [inline plsc/entry.py]: 这个地方我想起了一个问题：当前主干的nccl控制类是全局单例，每次run一个startup program中的init就会新建一个ring，所以这个地方这样使用不太符合预期。能否全局run startup，然后在这个地方只run broadcast 或者 allreduce？

Comment (sandyhouse) [reply plsc/entry.py]: run startup once for _barrier

**Pull request closed without merging**